### PR TITLE
Reorganize code

### PR DESCRIPTION
**Pull request by hunt0r**
In the process of redesigning following #45 , I first did essentially trivial re-organization and comment changes.  Since most of the code was touched, but each touch was pretty trivial, **I recommend you review commit-by-commit if possible.** That will make reviewing much easier.

P.S. I hope I used the "flatten" term correctly... when I say I "flatten" a function, I just mean that I copy-and-paste the function's code to the spot where it was called.

### CODE REVIEW

**Comment by Georacer** [comment]
Ooookay. I took a quick glance on the commits. I will be a bit vocal about it. Bare with me and let's try to have a proper debate.

In the first years of my EE diploma we had about 3 or 4 programming courses. I don't remember if I was taught this or I picked it up somewhere else, but the No1 advice and guideline I have when programming is
**Any function should ideally be able to be printed in one sheet of paper**

What this means, essentially, is that any program should be made up from logically well-defined, bitesize parts, with clear dependencies between them.
Each time you define a function, not only you isolate its code withing its namespace, but you impose a logical separation of functionality.

This is useful twofold:
1. You can quickly re-use that functionality in the future, if needed
2. Perhaps most importantly, future readers (including future you) will be able to identify and summarize the functionality that a function serves and be able to read the code in abstraction, instead of getting bogged down from line to line, something comments can't always provide; the barrier/delimiter of a function call is much stronger than any comment.

In conclusion, I am against any flattening of functions. Traditionally, flattening is used in programming to squeeze as much memory and execution speed as possible, but this is the least of our concerns at this point, especially if we hope to collaborate with 3rd parties.

Counter arguments?

**Comment by hunt0r** [comment]
I've never had coding-style training, so **I'm willing to adopt any guidelines**. I'll explain my thinking below, but I am happy to adopt whatever is tried-and-true.

My working idea is that a function should be used primarily (exclusively?) to eliminate duplicate code. If a logical-chunk is only used a single place, why not cut-and-paste it ("flatten" it) into that specific spot? If it's used more than once, in separate places, it should be functionalized. Likewise, if in writing new code, you want to copy-and-paste from somewhere else... probably time to functionalize.

In the same vein, if two similar logical-chunks of code exist, it is an improvement to formulate a single, general function which can handle both cases. (often a simple wrapper can translate between the specific and general forms) This can lead to abstract (hard-to-understand) code, but it's usually worth it. Comments can help with ambiguity/difficulty.

I also subscribe to an "evolving code" philosophy. That is: don't sweat too much about style/implementation decisions on the first go-around, but just get the code working. Don't worry much about picking the "best function/variable name." As the code develops, you'll learn more and have a better vision of the end product... and this allows re-organization for clarity. In my eyes, our code could benefit from some re-organization.

You can also see my coding-style is comment-heavy. I tend to prefer very verbose (often redundant) comments to crafting "self-commenting" code.


So: That's the counter-argument, for discussion's sake. But a big part of this project is me learning how to do collaborative coding, so I want to adopt tried-and-true best practices. If you recommend that we go with a function-fits-on-one-page style, say the word, and I'm in agreement. (I think I can re-functionalize this, or it's derivative which I'm working on now, pretty quickly.)

**Comment by Georacer** [comment]
What I definitely don't want is to come across as an authority in programming; I definitely am not. So I'll try to find a couple of books to support my position.

However, I'm leaving today for Easter holidays and will be back on the 24th. Can we postpone this conversation until then?

Happy Easter!

**Comment by hunt0r** [comment]
Sure, no problem. Have a great holiday!

**Comment by Georacer** [comment]
I'm back! Let's get to business! (I'm also answering to your post in #45)

You say:
>The key principle is "DRY: Don't Repeat Yourself" So if something is "repeated" in the code, it exists twice and should be condensed into a single source. (e.g. made into a separate function)

100% agree.
>Tracer bullet code: Often when designing/developing a program, it's best to code something, see how it works, then re-code (or trash it) as you discover better your use-cases and functionality needs.

Yes, we don't need to sweat form too much when in the first steps of prototyping.
>Re-factoring code (what I've been calling re-organizing) is a common and appropriate part of the coding cycle/process. Do it as often as you think it needs to be done.

Again, 100%. Re-write it as if you add strokes to a painting, to make it better and more beautiful.

So, you asked me
>If a logical-chunk is only used a single place, why not cut-and-paste it ("flatten" it) into that specific spot?

and I promised to come up with counter arguments. I would like to re-phrase it as "Why we shouldn't flatten existing functions" **not** as "Why you should always write code in functions".
I couldn't reach official textbooks on the topics.
I asked friends who work as professional programmers and they agreed with me but couldn't point to any valid sources. They provided some arguments which I expand on, with the help of some links, for **clarity** not **authority**.

Okay, the [first one](http://www.cs.utah.edu/~germain/PPS/Topics/functions.html) may carry some authority.
>Why do we Write Functions?
>1. They allow us to conceive of our program as a bunch of sub-steps. (Each sub-step can be its own function. When any program seems too hard, just break the overall program into sub-steps!)

This isn't really our point of debate.
>2. They allow us to reuse code instead of rewriting it.

This isn't it either, but having a function already there allows you to reuse code readily.
>3. Functions allow us to keep our variable namespace clean (local variables only "live" as long as the function does). In other words, function_1 can use a variable called i, and function_2 can also use a variable called i and there is no confusion. Each variable i only exists when the computer is executing the given function.

This is important and helps you write code without worrying about its context.
>4. Functions allow us to test small parts of our program in isolation from the rest. This is especially true in interpreted langaues, such as Matlab, but can be useful in C, Java, ActionScript, etc.

This is also important and is mandatory if we need to make unit tests of our code at some point.

Another topic touched [here](https://www.cs.utah.edu/~zachary/computing/lessons/uces-10/uces-10/node11.html) is _abstraction_.
>If you just want to use the function in your program, you don't have to know how it works inside! You don't have to understand anything about what goes on inside the function.

In other words, each time we re-read our code to make edits, we don't have to get through and over the minutia of what each function does. If it has a telling name and description, we can say "aha, I understand why this is here and what it's supposed to do" and go on. It makes code more readable.

So, [what makes a good function](https://nicercode.github.io/guides/functions/)?
1. It's short
2. It performs a single operation
3. It uses an intuitive naming scheme and interface

All of the above are re-phrased in [this discussion](https://softwareengineering.stackexchange.com/a/210405) where [this book](https://www.amazon.com/dp/0132350882/?tag=stackoverfl08-20) is proposed. I have not read it.

To sum up, I don't blame you for not writing functions every 100 lines of code. Far from it, that would be unproductive and constraining.
What I'm saying is that it's rather silly to "flatten" functions when you have already written them, since they provide with so much more than code reusability.

I'd be interested to hear what your acquaintance thinks of this.
In conclusion, all I ask is that existing functions are not "flattened".

**Comment by hunt0r** [comment]
Thanks for all this! I was neglecting many aspects of functions in my process.

I agree that flattening all functions (like this PR) isn't our best path forward. In fact, as I did it locally, I found that I later wanted to re-functionalize some of them identically to their original form!

In the break, I've done some thinking about our code design. I'll post about it in the code re-organization discussion #45.

In the interim, how about we reject this PR, and we can reconsider re-factoring PR's in the future as they come up?

**Comment by Georacer** [comment]
Fine by me! I'm glad you found my post useful.

Consider reading up on git's _cherry-pick_ functionality. It will save you a lot of time, picking only a subset of this PR's commits to re-submit for a new PR.
